packages/completer/src/widget.ts

```diff
@@ -564,85 +751,149 @@ export class Completer extends Widget {
     docPanel.textContent = '';
-    docPanel.appendChild(this._createLoadingBar());
+    const loadingIndicator = this._renderer.createLoadingDocsIndicator
+      ? this._renderer.createLoadingDocsIndicator()
+      : this._defaultRenderer.createLoadingDocsIndicator();
+    docPanel.appendChild(loadingIndicator);
+
     resolvedItem
       .then(activeItem => {
         if (!activeItem) {
           return;
         }
+        if (!docPanel) {
+          return;
+        }
         if (activeItem.documentation) {
-          let node: HTMLElement;
-          const nodeRenderer =
-            this._renderer.createDocumentationNode ??
-            Completer.getDefaultRenderer(this.sanitizer)
-              .createDocumentationNode;
-          node = nodeRenderer(activeItem);
-          docPanel!.textContent = '';
-          docPanel!.appendChild(node);
-          docPanel!.setAttribute('style', '');
+          const node = this._renderer.createDocumentationNode
+            ? this._renderer.createDocumentationNode(activeItem)
+            : this._defaultRenderer.createDocumentationNode(activeItem);
+          docPanel.textContent = '';
+          docPanel.appendChild(node);
         } else {
-          docPanel!.setAttribute('style', 'display:none');
+          this._toggleDocPanel(false);
         }
       })
       .catch(e => console.error(e));
   }
 
+  private _toggleDocPanel(show: boolean): void {
+    let docPanel = this._docPanel;
+    if (!docPanel) {
+      return;
+    }
+    if (show) {
+      if (this._docPanelExpanded) {
+        return;
+      }
+      docPanel.style.display = '';
+      this._docPanelExpanded = true;
+    } else {
+      if (!this._docPanelExpanded) {
+        return;
+      }
+      docPanel.style.display = 'none';
+      this._docPanelExpanded = false;
+    }
+    const sizeCache = this._sizeCache;
+    if (sizeCache) {
+      sizeCache.width += this._docPanelWidth * (show ? +1 : -1);
+      if (!this._geometryLock) {
+        this._setGeometry();
+      }
+    }
+  }
+
   private _activeIndex = 0;
   private _editor: CodeEditor.IEditor | null | undefined = null;
   private _model: Completer.IModel | null = null;
   private _renderer: Completer.IRenderer;
-  private _resetFlag = false;
+  private _defaultRenderer: Completer.Renderer;
   private _selected = new Signal<this, string>(this);
   private _visibilityChanged = new Signal<this, void>(this);
   private _indexChanged = new Signal<this, number>(this);
   private _lastSubsetMatch: string = '';
   private _showDoc: boolean;
+  private _sizeCache: Private.IDimensionsCache | undefined;
+
+  /**
+   * The maximum height of a completer widget.
+   */
+  private _maxHeight: number;
+
+  /**
+   * The minimum height of a completer widget.
+   */
+  private _minHeight: number;
+
+  private _scrollbarWidth: number;
+  private _docPanelWidth: number;
+  private _docPanel: HTMLElement | undefined;
+  private _geometryLock = false;
+  private _geometryCounter: number = 0;
```

Comment:
Could you add a docstring for the `_geometryCounter` variable?

Reply:
```suggestion
  /**
   * Increasing this counter invalidates previous request to save geometry cache in animation callback.
   */
  private _geometryCounter: number = 0;
```